Add new LOG class

MatthewScholefield opened the pull request:
This brings in a new way to log. Rather than putting `getLogger(__name__)` at the top of every file, you simply log with `LOG.debug('message')`. Here's an example:

```Python
from mycroft.util.log import LOG

def run():
    LOG.debug("Here's a debug message")
    LOG('CUSTOM_NAME').info('Another message')
    custom_logger = LOG.create_logger('MY_NAME')
    custom_logger.warn('warning?!')

run()
```

Output:
```Bash
14:03:44.061 - __main__:run:6 - DEBUG - Here's a debug message
14:03:44.061 - CUSTOM_NAME - INFO - Another message
14:03:44.061 - MY_NAME - WARNING - warning?!
```

As you can see, the default format is `module:function:line_number`. Since this isn't necessarily a high priority, I'm not assigning anyone yet.

As a side effect of touching all the import lines, Codacy complained of all the unused imports that already existed in the code. Because of this, this PR also removes all unused imports from the code base.

forslund commented on mycroft/util/log.py:
```diff
@@ -14,33 +14,116 @@
+    def init(cls):
+        sys_config = '/etc/mycroft/mycroft.conf'
+        level_name = load_commented_json(sys_config)['log_level'] \
+            if isfile(sys_config) else 'DEBUG'
```
This is a personal opinion, I find the ternary operation a bit confusing when split on different lines like this.

forslund commented on mycroft/util/log.py:
```diff
@@ -14,33 +14,116 @@
+    def create_logger(cls, name):
+        l = logging.getLogger(name)
+        l.propagate = False
+        if not hasattr(cls, 'level'):
```
Also a personal preference, using hasattr makes reading the code a bit harder than

instanciating level to `None` and using

```python
     if not cls.level:
         ...
```

My personal opinion is that hasattr should mainly be used if there is uncertainty about what object is being operated on.

But personal opinion so feel free to ignore :)

MatthewScholefield replied:
Good idea! Yeah, hasattr is pretty ugly

forslund commented:
I'm not sure about how atomic python calls are, could there be some threading issues here since the class variables are basically global?

forslund commented on mycroft/util/log.py:
```diff
@@ -14,33 +14,116 @@
+    @classmethod
+    def init(cls):
+        sys_config = '/etc/mycroft/mycroft.conf'
+        level_name = load_commented_json(sys_config)['log_level'] \
```
I believe this will fail if a sys config exitst but lacks a `log_level` entry  should probably have `.get('log_level', 'info')` or similar.

MatthewScholefield replied:
Here's my take on the subject:

Do we want to support there being a malformed system config? I'm pretty sure if someone deleted that there's a ton of other places in the code it would fail. Since the config gets updated at the same time as the code, I would think we should be able to assume it to always be there. Also, in the case that in the future a developer moves the log_level config value but forgets to update it here, rather than subtly introducing a big it will error out at the exact spot in the code that needs to be fixed.

forslund replied:
I'm not sure if a missing entry is considered malformed.

The `/etc/mycroft/mycroft.conf` is not used in any setups based on the github repo and the current template for the file does not include the log_level parameter so it would break logging on most installations.

MatthewScholefield replied:
Yeah, you're right; I didn't realize it's only loading the system config. Will change.

forslund commented:
Regarding docstrings, you should be able to forward the docstrings from logging.Logger.* to your implementations either with a decorator or just something like  `method.__doc__ = logger.Logging.method.__doc__` after the method has been declared. Maybe prepend a sentence saying this call wraps a call to Logger.

MatthewScholefield commented:
I've added a unit test to verify there's no threading issues. I've also added the docstrings, to methods, but done dynamically to reduce code duplication.

forslund commented:
Cool beans, as no one ever said. Looks good, I'll test run it properly tomorrow and will merge if nothing unexpected pops up!

Good thing testing for threading issues. I haven't got a good grip on python threading, with C EVERYTHING is a risk for threading issues :)

forslund commented on mycroft/messagebus/service/main.py:
```diff
@@ -14,16 +14,13 @@
-
 import tornado.ioloop as ioloop
 import tornado.web as web
-import tornado.autoreload as autoreload
```
This is missing and hinders the messagebus service from starting

MatthewScholefield replied:
Good catch! It's strange that despite the alias, it still imports the original name. Either way, I've changed it to use the alias now (and changed `import mod.mod as mod` to `from mod import mod`).

forslund commented:
This may be a good idea, my main concern would be that it gives a false sense of that we actually have good tests. Current ~40% status only lie a little bit. That said just loading the code is also a valid test.

Shouldn't a linter catch these type of things?
